### DISCUSSION TITLE
Managing upgrades for multiple ArgoCD on single cluster.

### DISCUSSION THREAD

**Discussion by kevin-kho** [argoproj/argo-cd#14230] [Q&A]
My current ArgoCD architecture follows the "compromise" style where there are multiple ArgoCDs on a single cluster separated by namespace ([as seen here](https://akuity.io/blog/argo-cd-architectures-explained/#The-compromise---Instance-Per-Logical-Group)).

I have two ArgoCDs:

- ArgoCD non-HA for dev deployments onto cluster A
- ArgoCD HA for prod deployments onto cluster B,C,D

It looks like there's there's shared CRDs (App, AppSet, AppProject) installed cluster-wide.

Say there's a new ArgoCD version down the road , and App, AppSet, and AppProject CRDs changes. How would I upgrade ArgoCD non-HA first, without affecting ArgoCD HA?

Currently I am thinking of changing architecture completely for my dev deployments, I would change it to "separation of concerns" [as seen here](https://akuity.io/blog/argo-cd-architectures-explained/#The-separation-of-concern---Instance-Per-Cluster)

for the ArgoCD dev, so it would be installed directly on target cluster A, and the only ArgoCD left on the management cluster is for prod deployments to cluster B,C,D.

Any thoughts or suggestions?

**Comment by jannfis** [comment]
We take great care that CRDs are backwards-compatible across minor releases. We do add new fields occasionally, but do not remove existing fields or change their semantics (unless absolutely necessary due to i.e. a security fix). Older versions should always be be working with a newer version of the CRD.

So a rolling upgrade of your Argo CD instances (with the first one upgrading the CRDs) should not impose a problem.

There are other pros and cons to the various architectures though, not related to upgrades. Obviously, it could make sense to have a testing or dev Argo CD instance also on your testing or dev cluster.
